### PR TITLE
Chenyan98/rsa

### PR DESCRIPTION
**Pull request by ChenYan98**
Create new PR rebase against dev

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs [hunk 1]
```diff
@@ -144,6 +144,46 @@ private bool IsSupportedKeyWrapAlgorithm(string algorithm, SecurityKey key)
...
+                    else
+                    {
+                        if (x509Key.PublicKey as RSACryptoServiceProvider != null)
+                            return true;
+                    }
+
+                    return false;
+                }
+
+                var jsonWebKey = key as JsonWebKey;
```
Comment (polita) [inline src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs]: Don't you need to check if key != null before using it?

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs [hunk 2]
```diff
@@ -111,6 +111,8 @@ internal static class LogMessages
         internal const string IDX10668 = "IDX10668: Unable to create '{0}', algorithm '{1}'; key: '{2}' is not supported.";
         internal const string IDX10669 = "IDX10669: Failed to create symmetric algorithm.";
         internal const string IDX10670 = "IDX10670: The lengths of the two byte arrays do not match. The first one has: '{0}' bytes, the second one has: '{1}' bytes.";
+        internal const string IDX10671 = "IDX10671: Unable to create the RsaKeyWrapProvider.\n\nKeyWrapAlgorithm: '{0}', SecurityKey: '{1}'\n is not supported.";
```
Comment (polita) [inline src/Microsoft.IdentityModel.Tokens/LogMessages.cs]: Remove extra \n.

File: src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs [hunk 3]
```diff
@@ -144,6 +144,46 @@ private bool IsSupportedKeyWrapAlgorithm(string algorithm, SecurityKey key)
...
+                return false;
+
+            if (string.IsNullOrEmpty(algorithm))
+                return false;
+
+            if (algorithm.Equals(SecurityAlgorithms.RsaPKCS1, StringComparison.Ordinal)
+                || algorithm.Equals(SecurityAlgorithms.RsaOAEP, StringComparison.Ordinal)
+                || algorithm.Equals(SecurityAlgorithms.RsaOAEP256, StringComparison.Ordinal))
+            {
+                if (key as RsaSecurityKey != null)
```
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs]: Modified

File: src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs (hunk 1 above)
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs]: Check jsonWebKey != null

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs (hunk 2 above)
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Tokens/LogMessages.cs]: Modified

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        /// </summary>
+        /// <param name="disposing">true, if called from Dispose(), false, if invoked inside a finalizer.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+#if NETSTANDARD1_4
+                    if (_rsa != null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: should not dispose unless _rsa was created. See https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L228

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+        }
+
+        /// <summary>
+        /// Checks if an algorithm is supported.
+        /// </summary>
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The KeyWrap algorithm to apply.</param>
+        /// <param name = "willDecrypt"> Whether this <see cref = "RsaKeyWrapProvider" /> is required to create decrypts then set this to true.</param>
+        /// <returns>true if the algorithm is supported; otherwise, false.</returns>
+        protected virtual bool IsSupportedAlgorithm(SecurityKey key, string algorithm, bool willDecrypt)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: what is 'willDecrypt' used for?

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+            RsaSecurityKey rsaKey = key as RsaSecurityKey;
+            if (rsaKey != null)
+            {
+                if (rsaKey.Rsa != null)
+                {
+                    _rsa = rsaKey.Rsa;
+                    return;
+                }
+
+                _rsa = RSA.Create();
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: In this case we need to dispose '_rsa'. see: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L228

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+                    return;
+                }
+            }
+
+            X509SecurityKey x509Key = key as X509SecurityKey;
+            if (x509Key != null)
+            {
+                if (willDecrypt)
+                {
+                    RSACryptoServiceProvider rsaCsp = x509Key.PrivateKey as RSACryptoServiceProvider;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: We will need to account for RSACng on windows.

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs (hunk 5 above)
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: 'willDecrypt' will be set to true if the RsaKeyWrapProvider instance used to decrypt later. If SecurityKey is X509SecurityKey, then check PrivateKey for decrypt; check PublicKey for encrypt.

File: src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs [hunk 8]
```diff
@@ -144,6 +144,46 @@ private bool IsSupportedKeyWrapAlgorithm(string algorithm, SecurityKey key)
             return false;
         }
 
+        private bool IsSupportedRsaKeyWrapAlgorithm(string algorithm, SecurityKey key, bool willDecrypt)
```
Comment (ChenYan98) [inline src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs]: Removed

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+        /// </summary>
+        public RsaKeyWrapProvider(SecurityKey key, string algorithm, bool willDecrypt)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (string.IsNullOrEmpty(algorithm))
+                throw LogHelper.LogArgumentNullException(nameof(algorithm));
+
+            if (key.KeySize < 2048)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: 2048 is min for wrapping, 1024 for unwrapping.

see: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L82

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+
+            if (!IsSupportedAlgorithm(key, algorithm))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10671, algorithm, key)));
+
+            Algorithm = algorithm;
+            Key = key;
+
+            ResolveRsaAlgorithm(key, algorithm, willDecrypt);
+#if NETSTANDARD1_4
+            if (_rsa == null && _rsaCryptoServiceProviderProxy == null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: _rsaCryptoServiceProviderProxy should only be in scope fro NET451. See: comments on https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/559

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+        /// </summary>
+        /// <remarks>This is null by default. This can be used by runtimes or for extensibility scenarios.</remarks>
+        public string Context { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="SecurityKey"/> that is being used.
+        /// </summary>
+        public SecurityKey Key { get; private set; }
+
+        private RSAParameters CreateRsaParametersFromJsonWebKey(JsonWebKey webKey, bool isPrivate)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: We should consolidate this into a function that can be accessed from different locations. A similar function seems to be found here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L383

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+                {
+#if NETSTANDARD1_4
+                    if (_rsa != null && _disposeRsa)
+                        _rsa.Dispose();
+#else
+                    if (_rsaCryptoServiceProvider != null && _disposeRsa)
+                        _rsaCryptoServiceProvider.Dispose();
+#endif
+
+                    if (_rsaCryptoServiceProviderProxy != null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: See comments in: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/559

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Checks if an algorithm is supported.
+        /// </summary>
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The KeyWrap algorithm to apply.</param>
+        /// <returns>true if the algorithm is supported; otherwise, false.</returns>
+        protected virtual bool IsSupportedAlgorithm(SecurityKey key, string algorithm)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: We don't check key size.

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+
+#if NETSTANDARD1_4
+        /// <summary>
+        /// Initialize RSA algorithm.
+        /// </summary>
+        /// <param name="key">The <see cref="SecurityKey"/> that will be used for crypto operations.</param>
+        /// <param name="algorithm">The RSA KeyWrap algorithm to apply.</param>
+        /// <param name="willDecrypt">Whether this <see cref="RsaKeyWrapProvider"/> is required to create decrypts then set this to true.</param>
+        /// <exception cref="ArgumentOutOfRangeException">The <see cref="SecurityKey"/> is not supported.</exception>
+        protected virtual void ResolveRsaAlgorithm(SecurityKey key, string algorithm, bool willDecrypt)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: This is very similar to: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L319
Can we have a single function?

File: src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs [hunk 15]
```diff
@@ -465,6 +512,51 @@ public virtual KeyedHashAlgorithm CreateKeyedHashAlgorithm(byte[] keyBytes, stri
             }
         }
 
+        private RsaKeyWrapProvider CreateRsaKeyWrapProvider(SecurityKey key, string algorithm, bool willDecrypt)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/CryptoProviderFactory.cs]: The operations on KeyWrap are Wrap and UnWrap. the parameter named willDecrypt seems incorrect.

File: src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs (hunk 9 above)
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs]: Just checked the spec, 2048 is required. So, leave this as is.

File: test/Microsoft.IdentityModel.Tokens.Tests/RsaKeyWrapProviderTests.cs [hunk 16]
```diff
@@ -0,0 +1,331 @@
+﻿//------------------------------------------------------------------------------
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tokens.Tests/RsaKeyWrapProviderTests.cs]: Please add references in: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/test/Microsoft.IdentityModel.Tokens.Tests/References.cs#L29
from: https://tools.ietf.org/html/rfc7516#appendix-A.1.2 and https://tools.ietf.org/html/rfc7516#appendix-A.3.3

then add tests into https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/test/Microsoft.IdentityModel.Tokens.Tests/ReferenceTests.cs#L36

Will need new tests for KW.

**Comment by brentschmaltz** [comment]
we are using keywrap for this set of changes

**Pull request closed without merging**